WIP PH5 data availability

derick-hess opened the pull request:
Creating data availability capabilities in PH5. The API now  supports getting extent and availability at the channel level.

Started to implement PH5 Availability client

nick-iris commented:
@derick-hess I'm trying to use the `get_availability` method to generate a tsindex database but get the following error:

```
        paths = get_experiment_paths()
        for path in paths:
            ph5 = ph5api.PH5(path=path, nickname="master.ph5")
            ph5.read_array_t_names()
            self.read_arrays(ph5)
            self.array_names = ph5.Array_t_names
            self.array_names.sort()
            for array_name in self.array_names:
                arraybyid = ph5.Array_t[array_name]['byid']
                arrayorder = ph5.Array_t[array_name]['order']
                for station in arrayorder:
                    station_list = arraybyid.get(station)
                    for deployment in station_list:
                        station_entry = station_list[deployment][0]
                        avail = ph5.get_availability(
                                    station_entry["das/serial_number_s"],
                                    station_entry["deploy_time/epoch_l"],
                                    station_entry["pickup_time/epoch_l"],
                                    station_entry["sample_rate_i"])
                        print(avail)
```

results in 

```
Traceback (most recent call last):
  File "PH5WebService/PH5Availability/loader/tsindex_loader.py", line 73, in <module>
    run_load_tsindex()
  File "PH5WebService/PH5Availability/loader/tsindex_loader.py", line 64, in run_load_tsindex
    PH5TSIndexLoader().load_tsindex()
  File "PH5WebService/PH5Availability/loader/tsindex_loader.py", line 52, in load_tsindex
    station_entry["pickup_time/epoch_l"])
  File "/Users/nick/Development/workspace/PH5/ph5/core/ph5api.py", line 1465, in get_availability
    start, stop = self.get_extent(das, start, end, component)
  File "/Users/nick/Development/workspace/PH5/ph5/core/ph5api.py", line 1408, in get_extent
    earliest_epoch = (float(new_das_t[0]['time/epoch_l']) +
IndexError: list index out of range
```

If no data availability information if found matching the request, I think a custom exception should be raised.

derick-hess commented:
@nick-iris I'll fix that and also look for any other error handling that needs to be added in the two new methods and have a pull request for you to test monday morning.

derick-hess commented on ph5/core/ph5api.py:
```diff
@@ -1397,15 +1397,31 @@ def receiver_cut(self, event_t_name, array_t, length):
 
     def get_extent(self, das, start=None, end=None, component=None):
```
Thanks Nick. I'll fix this. I think originally I was going to make component optional, but was looking at other methods in PH5 that extent needs and it would take more changes than I have time for now. I will change this to a parameter instead of a keyword argument.

nick-iris commented:
@derick-hess I wrote a program that loads a TSIndex database table using the [ph5.core.ph5api.PH5.get_availability](https://github.com/PIC-IRIS/PH5/blob/availability/ph5/core/ph5api.py#L1440) method. When [ph5.core.ph5api.PH5.get_availability](https://github.com/PIC-IRIS/PH5/blob/availability/ph5/core/ph5api.py#L1440)  is called repeatedly in a loop this method opens too many files and an exception is raised. Is there a reason the miniPH5 files opened by the ph5api are not closed until the entire [ph5.core.ph5api.PH5]() object is closed?

```
Traceback (most recent call last):
  File "PH5WebService/PH5Availability/loader/tsindex_loader.py", line 134, in <module>
  File "PH5WebService/PH5Availability/loader/tsindex_loader.py", line 125, in run_load_tsindex
  File "PH5WebService/PH5Availability/loader/tsindex_loader.py", line 73, in load_tsindex
  File "/Users/nick/Development/workspace/PH5/ph5/core/ph5api.py", line 1459, in get_availability
  File "/Users/nick/Development/workspace/PH5/ph5/core/ph5api.py", line 838, in read_das_t
  File "/Users/nick/Development/workspace/PH5/ph5/core/experiment.py", line 890, in setcurrent
  File "/Users/nick/anaconda/envs/ph5/lib/python2.7/site-packages/tables/link.py", line 396, in __call__
  File "/Users/nick/anaconda/envs/ph5/lib/python2.7/site-packages/tables/file.py", line 318, in open_file
  File "/Users/nick/anaconda/envs/ph5/lib/python2.7/site-packages/tables/file.py", line 784, in __init__
  File "tables/hdf5extension.pyx", line 488, in tables.hdf5extension.File._g_new (tables/hdf5extension.c:5584)
tables.exceptions.HDF5ExtError: HDF5 error back trace

  File "H5F.c", line 604, in H5Fopen
    unable to open file
  File "H5Fint.c", line 992, in H5F_open
    unable to open file: time = Wed Feb 20 19:45:33 2019
, name = '/hdf5-data2/PH5_Experiments/pn4/16-015/miniPH5_00245.ph5', tent_flags = 0
  File "H5FD.c", line 993, in H5FD_open
    open failed
  File "H5FDsec2.c", line 339, in H5FD_sec2_open
    unable to open file: name = '/hdf5-data2/PH5_Experiments/pn4/16-015/miniPH5_00245.ph5', errno = 24, error message = 'Too many open files', flags = 0, o_flags = 0

End of HDF5 error back trace

Unable to open/create file '/hdf5-data2/PH5_Experiments/pn4/16-015/miniPH5_00245.ph5'
```

derick-hess commented:
It should only open the files when it is being used. I think using the forget_das method in ph5api will close the minifile since it's not needing to access anything in the das anymore. I believe all the other program do this after they are done reading that das table. I'm not sure I can put this in teh availability or extent methods because that could prematurely close the das  for the greater program using it. It would probably be best to forget_das at the end of each iteration in your program. forget_das simply removes the das node from ph5 and therefore the minifile is no longer being read (unless their is another das in that minifile that is also open)

nick-iris commented:
I just tried using the `ph5availability` command on the same experiment and get the same error:

```
ph5availability -n master.ph5 -p /hdf5-data2/PH5_Experiments/pn4/16-015/ 

[2019-02-20 23:04:25,189] - ph5.clients.ph5availability - ERROR: HDF5 error back trace

  File "H5F.c", line 604, in H5Fopen
    unable to open file
  File "H5Fint.c", line 992, in H5F_open
    unable to open file: time = Wed Feb 20 23:04:25 2019
, name = '/hdf5-data2/PH5_Experiments/pn4/16-015/miniPH5_00263.ph5', tent_flags = 0
  File "H5FD.c", line 993, in H5FD_open
    open failed
  File "H5FDsec2.c", line 339, in H5FD_sec2_open
    unable to open file: name = '/hdf5-data2/PH5_Experiments/pn4/16-015/miniPH5_00263.ph5', errno = 24, error message = 'Too many open files', flags = 0, o_flags = 0

End of HDF5 error back trace
```

I tried adding a call to the `forget_das_t` method at the end of `get_availability` and it doesn't fix the issue. Any other ideas?

derick-hess commented:
I'll see what I can come up with. I have something almost exactly liek what you have. It loops through each station and checks availability. I'll test it on wavefields to see what happens.

Never seen that issue before even  when we wrote every bit of data from wavefileds using ph5toms

nick-iris commented:
Thanks and yes they are nearly identical. I'm also surprised that we haven't had this issue with `ph5toms` as it also appears to not close files until the end. I'm wondering if this could be part of the reason why `ph5toms` uses a more memory the longer it runs. Could it possibly be loading each das table into memory and never freeing any space?

Also it appears that each miniPH5 file contains a Receiver_g group containing Das_g groups. These appear to be identical to the Das_g groups stored in the master.ph5 file. Is there a reason they are in both the miniPH5_#####.ph5 files and master.ph5 file?

derick-hess commented:
There is no das_g in the master. That is just reading the data from inside the mini_ph5file as an external reference. In the master it has das_g_xxxxx and the contents of das_g_xxxxx both the das_table and arrays of data are actually stored in the mini files.

Last time i ran ph5toms memory tests, it jumps up but then pretty much levels out

derick-hess commented:
pytables supports removing a node from memory so in the morning I will try adding that to forget_das_t as well to see if that helps.

nick-iris commented:
Unmounting the ExternalLink at the end of the `get_availability` method fixes the problem:

```
node = self.ph5_g_receivers.getdas_g(das)
node.umount()
```

derick-hess commented:
Thanks I thought it might. I'm out sick today, but I will go through ph5api and experiment and go through to make sure nodes are getting freed and arrays are properly deleted as soon they are no longer needed.

derick-hess commented:
I have modified ph5api.forget_das() to unmount the node after use. I have added this to the ph5api.cut() method as well as ph5api.get_availability() and ph5api.get_extent()

I have added a framework for ph5 availability package/application. Below is Nick's table form the issue for reference. Modified for PH5. 

Method | Return Type | Description
-- | -- | --
get_nslc(station, location, channel, starttime, endtime) | list(tuple) | Return a list of tuples [(net, sta, loc, cha),...] containing information on what streams are included in PH5 Archive.
get_availability_extent(station, location, channel, starttime, endtime) | list(tuple) | Return a list of tuples [(net, sta, loc, cha, earliest, latest)] containing data extent info for time series included PH5 Archive
get_availability(station, location, channel, starttime, endtime,include_sample_rate=False) | list(tuple) | Return a list of tuples [(net, sta, loc, cha, start, end),...] containing data availability info for time series included in the PH5 archive. If include_sample_rate=True, then a tuple containing the sample rate [(net, sta, loc, cha, start, end, sample_rate),...] is returned.
get_availability_percentage(station, location, channel, starttime, endtime) | 2-tuple of percentage of available data (`0.0` to `1.0`) and number of gaps/overlaps. | Get percentage of available data for a specified station, location, channel, starttime, endtime combination.
has_data(station, location, channel, starttime, endtime) | boolean | Check if specified stream has any data using the PH5 Archive.

derick-hess commented on ph5/core/ph5api.py:
```diff
@@ -1419,9 +1438,61 @@ def get_extent(self, das, start=None, end=None, component=None):
+        '''
+        if not component:
+            raise ValueError("Comonent required for get_avilability")
+        if not sample_rate:
```
I've been considering this for other reasons as well. The das_t only has a start time and number of samples in the window. It would make things way easier if it also had and end time available. Then there would be no need to do these calculations unless it's an old experiment and the end time isn't there.

For now until we get approval from standing committee to add new fields to PH5 we can do the same thing and have 0 sample rate channel/component_num set as a negative number. If it is negative then the sample_count is the length of the window in microseconds instead of number of samples  and the latest_epoch can still be calculated without a divide by zero error

derick-hess replied:
Actually for now i'm totally over thinking this. ALL the miniSEED logs I have found are just timestamped with each entry and length of 0 so every window has no length just a start time. So until approval for new end time in das_t, then if sample_rate is 0 end is the same as start so no need to do the division in that case just a check if sample rate is 0 or not

derick-hess commented:
@damhuonglan  I have fixed the 0 sample rate issues in the miniseedtoph5 PR.  Don;t worry about those.
I think the table above defines everything pretty well. ph5API contains get_extent and get_availability at the channel level that can be used to help build this.

derick-hess commented:
1) You will have to pick one way or another. If there is less than 100% data, then there must be gaps otherwise you would have 100% data coverage.  IF you are reporting that test as less than 100% then gaps must exist at the start and end of the specified time range.

2) LOG and text channels are handled special. The sample_rate is 0 so by definitions that means there are no samples.  If sample_rate is 0, then you need to check the data array specified in that das table entry to see if it contains data.

derick-hess commented:
@damhuonglan I'm starting to write the get_availability_extent tests and it fails immediately.
ret = self.availability.get_availability_extent()

ph5api.get_extent() raises an error:
"if start or end, both are required"

I have't provided a start time or an end time. When your code calls get_extent it must give both a start time and end time or give no start time and no end time.

derick-hess commented:
@damhuonglan I just pushed a few simple test for get_availability_extent. I will push more tests later this afternoon, but these should get you started on get_availability_extent

derick-hess commented:
@damhuonglan similarly you aren't giving ph5api.get_availability a required input which is the sample_rate
it raises an error 'Sample rate required for get_availability'

nick-iris commented:
Is it possible to get this synced with the master branch so that we have the test data?

I'm running the get_availability method against the entire archive and it will occasionally break, so I would like to make sure that the unit test are still passing. I've fixed a few bugs.

derick-hess commented:
It already is. From the main PH5 directory run `bash create_ph5.sh` and it will create a test ph5 archive in ph5/test_data/ph5.   Eventually I need to add some more test data, but for now this does okay. It has a node, rt130, miniseed, and rt125 laoded in. A LOG channel is included and the rt125 data has 8 gaps to help test availability.

because of the way I had to do the paths. You need to run test_ph5availability.py from the main PH5 directory like this `python ph5/clients/tests/test_ph5availability.py`

the ph5api tests also test the extent and availability methods that run on a channel level at in the api

derick-hess commented:
I went ahead and fixed the conflicts and merged master in to availability fully

derick-hess commented:
@nick-iris feel free to add more tests in test_ph5api for get_availability and get_extent. If you see a type of data we need to add to the data set let me know and I can create it

derick-hess commented:
@damhuonglan  That is tested for in the ph5api test and returns correct lines 1020-1038. Correct values are returned in both cases.

https://github.com/PIC-IRIS/PH5/blob/796605a6e13b63ded14f8e488596f90ce0444290/ph5/core/tests/test_ph5api.py#L1020-L1038

derick-hess commented:
You can remove that line as long as all the tests still pass

derick-hess commented on ph5/clients/ph5availability.py:
```diff
@@ -247,20 +273,61 @@ def get_availability_extent(self, station='*', location='*',
 
         return availability_extents
 
+    def get_das_avail(self, das_t, samplerate, earliest, latest, start, end):
```
a get_availability at the das level already exists in ph5api. Cutting data is too slow. There is no need to actually grab data.  This could lead to memory issues on large requests and make things too slow to work

derick-hess commented:
@damhuonglan so far this looks good. Tested on local machine and everything seems to work. Travis CI the service that tests the code online is having issues so it should eventually run and output it's results.  I will add a few more unit tests to test_ph5availability to cover more code as well

derick-hess commented:
@damhuonglan here are some resources for the file format output

http://www.iris.washington.edu/bud_stuff/goat/syncformat.html

http://geows.ds.iris.edu/documents/GeoCSV.pdf

https://cloud.passcal.nmt.edu/index.php/s/QsCn4wkNHNoFprE

derick-hess commented:
To make them more reusable I would put the file output code in ph5utils to make it more easily usable by other programs.

nick-iris commented on ph5/clients/ph5availability.py:
```diff
@@ -129,23 +135,40 @@ def get_array_order_id(self, array_name):
+                    LOGGER.warning("No Das table found for %s in range %s - %s"
+                                   % (das, start, end))
+                return -1
+            Das_t = deepcopy(self.ph5.Das_t[das]['rows'])
```
Is this `deepcopy` call really necessary? This seems expensive.

derick-hess replied:
I don't think deepcopy is needed here.

damhuonglan replied:
I will remove that. I was afraid the normal copy doesn't work after forget_das_t but actually it still work after removing deepcopy.

nick-iris replied:
@damhuonglan Are you sure a copy is necessary at all? I think it should be avoided completely if possible.

derick-hess replied:
`Das_t = self.ph5.Das_t[das]['rows']` should be okay. Lists are passed by reference so no copying is being done. But I think you are using self.ph5.forget_das_t(das) a bit early. You move that to the end of the method for safety.

nick-iris commented:
Something is not right here.

The following query returns no availability results:
```
ph5availability -a 2 -n master.ph5 -p /hdf5-data2/PH5_Experiments/pn4/15-016/ --station=1002 --channel=DPZ -s=2015-06-29T06:25:00.019999 -t=2015-06-29T06:26:00.019999
```

But ph5toms returns data for the same query:
```
ph5toms -p /hdf5-data2/PH5_Experiments/pn4/15-016/ -n master.ph5 --station=1002 --channel=DPZ -s=2015-06-29T06:25:00.019999 -t=2015-06-29T06:26:00.019999
```

I'm beginning to do some testing since I now have an availability web service in beta which sources its information from this availability API.

derick-hess commented:
Thanks Nick I will start doing tests on this and the channel level built in to the API this afternoon as well

derick-hess commented:
@nick-iris @damhuonglan  I think I see the issue. The underlying seem okay but the first issue I see is when the time arguments are being collected from the command line they aren't being properly converted to epochs

derick-hess commented:
also all of these extra methods that have been added as well as the main will need to have unit tests done for them.

 get_time_das_t, analyze_args, get_channel, get_slc_info, get_start, get_end, get_sample_rate, get_sampleNos_gapOverlap, process_all, get_args, main

I will update the get_args method for Lan a bit to make it unit test friendly.

nick-iris commented:
@derick-hess I have another program that calls the `get_availability` method directly and it has the same problem. For this reason I think there is probably another problem besides times not being converted properly on the command line.

derick-hess commented:
I think you just have a typo, or ph5toms has an issue. The time range in your query is a year before there is any data.

When I adjust it to 2016 ph5api.get_availability() returns
[(250.0, 1467179948.888, 1467181748.888)]

nick-iris commented:
@derick-hess According to the station service, the year is correct:

https://service.iris.edu/ph5ws/station/1/query?level=channel&format=geocsv&reportnum=15-016&station=1002

```
#dataset: GeoCSV 2.0
#delimiter: |
#field_unit: unitless | unitless | unitless | unitless | degrees_north | degrees_east | meters | meters | degrees | degrees | unitless | factor | hertz | unitless | hertz | ISO_8601 | ISO_8601 | unitless | unitless | unitless | unitless
#field_type: string | string | string | string | float | float | float | float | float | float | string | float | float | string | float | datetime | datetime | string | string | string | string
Network|Station|Location|Channel|Latitude|Longitude|Elevation|Depth|Azimuth|Dip|SensorDescription|Scale|ScaleFreq|ScaleUnits|SampleRate|StartTime|EndTime|ReportNumber|Array|Receiver|Component
ZI|1002||DPZ|36.0859|-75.70414|2.75|0.0|0.0|90.0|Geo Space/OYO GS-11D/REF TEK RT 125 & 125A|1626795077.54|62.5|M/S|250.0|2015-06-29T02:00:00|2015-07-01T03:00:00|15-016|001|1002|1
```

derick-hess commented:
Oops sorry that was my fault. I'm looking at 16-015... I have 15-016 data as well here at teh PIC. I'll check it out after lunch

derick-hess commented:
@nick-iris I get data back for that request when called directly.

when i call ph5api.get_availability() directly with this:
`ph5API_object.get_availability('11177', 250, 1, 1435559100.019999, 1435559160.019999)`
output is: `[(250.0,` 1435557600.0, 1435561200.0)]`
This is because when no gaps are detected it gives extent information...this behavior can be changed if incorrect

calling ph5availability.get_availability() directly with this:
`availability.get_availability('1002', '', 'DPZ', 1435559100.019999, 1435559160.019999)`
output is `[('1002', '', 'DPZ', 1435559100.019999, 1435559160.019999)]`

derick-hess commented:
I haven't tested passing a UTCDateTime from ObsPy yet though. According to the method description that should also be supported

derick-hess commented:
I do see an issue with get_availability though. Working on a fix.

derick-hess commented:
okay pushed a fix so gaps should be accounted for correctly again in ph5api.get_availability()
Also went ahead and removed deepcopy

derick-hess commented:
I will write unit tests for main() before I leave for EGU to help with testing. Lan, you will need to write the unit tests for the rest of the added methods

nick-iris commented:
@derick-hess Currently iso8601 dates are not supported by `ph5availability`, only PASSCAL epoch dates in the format `YYYY:DOY:HH:MM:SS.ss` and epoch timestamps.

I'd like to add support for iso8601 dates. It useful to be able to take web service arguments and run them against the PH5 terminal commands.

derick-hess commented:
Thanks Nick. That should definitely be be supported. @damhuonglan command line time arguments for time should match ph5toms. It's okay to also support epoch times but at a minimum PASSCAL time and iso8601 (yyyy-mm-ddThh:mm:ss.ffffff) should be supported like in the other ph5 command line clients

derick-hess commented:
@damhuonglan I would like to have this ready in the coming week so we can get other things merged and a release out.

nick-iris commented:
I added some improved argument parsing methods that can be used for all clients. Additionally the program now supports iso dates.

I think the outputs need to be presented in a different format. Currently a bunch of information is returned with no context. For example, the `get_availability` method should return a list of tuples but the `availability` output (option --avail=2) should probably return text in the format matching the availability web service.

derick-hess commented:
@nick-iris that's the plan. Command line output is supposed to support text, sync, geocsv, json and request.

I'll take a look at the args stuff. I might have to change it a little bit in order to be able to pass arguments correctly so that unit tests can handle the main() method as well.

derick-hess commented:
I hand created a geocsv and text output for all channels of the test data set and added that to the unit tests as well as a few test for has_data

nick-iris commented:
@derick-hess  and @damhuonglan, what is remaining on this PR before we can merge?

derick-hess commented:
@nick-iris I was out yesterday, but I am starting testing on this PR today to see where it stands

derick-hess commented:
@damhuonglan this is pretty close. IT still has some flake8 errors that need to be fixed

derick-hess commented:
@damhuonglan can you get the errors fixed? This needs to get merged soon

damhuonglan commented:
I fixed the error but cannot commit them because of the error: Committing is not possible because you have unmerged files.
I have to fix those errors first.

derick-hess commented:
You need to make sure you always pull the latest.

nick-iris commented:
I found a few bugs that I am working on addressing now, but this is very close to being ready.

I'm wondering if we should eliminate the `has_data` and `availability_percentage` output options. Looking at their outputs, they don't seem particularly useful for the client. Thoughts?

derick-hess commented:
Yeah those two don't really need to be there. We can finish this PR like it is and when Lan gets back from vacation she can add the last of the client stuff (outputting in various formats) in a new PR

nick-iris commented:
If there are no objections I'd like to merge this branch. @damhuonglan @derick-hess Is this okay with you?
